# Reading the clean dataset in a dataframe

In [10]:
 import pandas as pd
df1 = pd.read_csv("E:\Machine Learning\Crime Prediction Data2\Crime Prediction Data\communities-crime-clean.csv")
df1

state        communityname  fold  population  householdsize  \
0         1        Alabastercity     7        0.01           0.61   
1         1    AlexanderCitycity    10        0.01           0.41   
2         1         Annistoncity     3        0.03           0.34   
3         1           Athenscity     8        0.01           0.38   
4         1           Auburncity     1        0.04           0.37   
5         1         Bessemercity     6        0.04           0.44   
6         1       Birminghamcity     2        0.41           0.37   
7         1          Cullmancity     1        0.01           0.30   
8         1           Daphnecity     7        0.00           0.39   
9         1          Decaturcity    10        0.06           0.39   
10        1           Dothancity     4        0.07           0.41   
11        1       Enterprisecity     2        0.02           0.43   
12        1          Eufaulacity     5        0.01           0.46   
13        1        Fairfieldcity     1        0.00           0.45   
14        1         Florencecity     3        0.04           0.33   
15        1        FortPaynecity     2        0.00           0.35   
16        1          Gadsdencity     9        0.05           0.32   
17        1        Hartsellecity     4        0.00           0.43   
18        1         Homewoodcity     5        0.02           0.23   
19        1           Hoovercity    10        0.05           0.35   
20        1         Hueytowncity     4        0.01           0.44   
21        1       Huntsvillecity     1        0.24           0.37   
22        1     Jacksonvillecity     7        0.00           0.54   
23        1           Jaspercity     8        0.01           0.37   
24        1          Madisoncity     5        0.01           0.36   
25        1           Mobilecity     8        0.30           0.41   
26        1       Montgomerycity     2        0.28           0.45   
27        1    MountainBrookcity     8        0.02           0.37   
28        1        Northportcity     2        0.01           0.44   
29        1          Opelikacity     6        0.02           0.45   
...     ...                  ...   ...         ...            ...   
1963     55        NewBerlincity     4        0.04           0.55   
1964     55         OakCreekcity     6        0.02           0.49   
1965     55       Oconomowoccity     6        0.00           0.42   
1966     55         Onalaskacity     4        0.00           0.44   
1967     55          Oshkoshcity     7        0.07           0.42   
1968     55           Racinecity    10        0.12           0.44   
1969     55       RiverFallscity     3        0.00           0.67   
1970     55        Sheboygancity     8        0.06           0.37   
1971     55     Shorewoodvillage     6        0.01           0.18   
1972     55   SouthMilwaukeecity     9        0.02           0.39   
1973     55     StevensPointcity     4        0.02           0.51   
1974     55       SunPrairiecity     2        0.01           0.48   
1975     55         Superiorcity     8        0.03           0.34   
1976     55        TwoRiverscity     6        0.00           0.37   
1977     55        Watertowncity     2        0.00           0.52   
1978     55         Waukeshacity     6        0.08           0.45   
1979     55           Wausaucity     7        0.04           0.37   
1980     55        Wauwatosacity     9        0.06           0.35   
1981     55        WestAlliscity     6        0.09           0.29   
1982     55         WestBendcity     3        0.02           0.49   
1983     55  WhitefishBayvillage     9        0.01           0.42   
1984     55       Whitewatercity     6        0.00           0.65   
1985     55  WisconsinRapidscity     8        0.01           0.32   
1986     56         Cheyennecity     4        0.06           0.34   
1987     56         Evanstoncity     6        0.00           0.65   
1988     56         Gillettecity     9        0.01           0.53   
1989

Create a new field “highCrime” which is true if the crime rate per capita (ViolentCrimesPerPop) is greater than 0.1, and false otherwise. What are the percentage of positive and negative instances in the dataset?

In [11]:
df1['highcrime'] = df1['ViolentCrimesPerPop'].map(lambda x : True if x >0.1 else False)
df1['highcrime'].head()


0    False
1     True
2     True
3     True
4     True
Name: highcrime, dtype: bool

In [6]:

print (pd.DataFrame({'Percentage': df1.groupby(('highcrime')).size() / len(df1)}))
 



           Percentage
highcrime            
False        0.372805
True         0.627195


In [12]:
df2 = df1.copy()
df2


state        communityname  fold  population  householdsize  \
0         1        Alabastercity     7        0.01           0.61   
1         1    AlexanderCitycity    10        0.01           0.41   
2         1         Annistoncity     3        0.03           0.34   
3         1           Athenscity     8        0.01           0.38   
4         1           Auburncity     1        0.04           0.37   
5         1         Bessemercity     6        0.04           0.44   
6         1       Birminghamcity     2        0.41           0.37   
7         1          Cullmancity     1        0.01           0.30   
8         1           Daphnecity     7        0.00           0.39   
9         1          Decaturcity    10        0.06           0.39   
10        1           Dothancity     4        0.07           0.41   
11        1       Enterprisecity     2        0.02           0.43   
12        1          Eufaulacity     5        0.01           0.46   
13        1        Fairfieldcity     1        0.00           0.45   
14        1         Florencecity     3        0.04           0.33   
15        1        FortPaynecity     2        0.00           0.35   
16        1          Gadsdencity     9        0.05           0.32   
17        1        Hartsellecity     4        0.00           0.43   
18        1         Homewoodcity     5        0.02           0.23   
19        1           Hoovercity    10        0.05           0.35   
20        1         Hueytowncity     4        0.01           0.44   
21        1       Huntsvillecity     1        0.24           0.37   
22        1     Jacksonvillecity     7        0.00           0.54   
23        1           Jaspercity     8        0.01           0.37   
24        1          Madisoncity     5        0.01           0.36   
25        1           Mobilecity     8        0.30           0.41   
26        1       Montgomerycity     2        0.28           0.45   
27        1    MountainBrookcity     8        0.02           0.37   
28        1        Northportcity     2        0.01           0.44   
29        1          Opelikacity     6        0.02           0.45   
...     ...                  ...   ...         ...            ...   
1963     55        NewBerlincity     4        0.04           0.55   
1964     55         OakCreekcity     6        0.02           0.49   
1965     55       Oconomowoccity     6        0.00           0.42   
1966     55         Onalaskacity     4        0.00           0.44   
1967     55          Oshkoshcity     7        0.07           0.42   
1968     55           Racinecity    10        0.12           0.44   
1969     55       RiverFallscity     3        0.00           0.67   
1970     55        Sheboygancity     8        0.06           0.37   
1971     55     Shorewoodvillage     6        0.01           0.18   
1972     55   SouthMilwaukeecity     9        0.02           0.39   
1973     55     StevensPointcity     4        0.02           0.51   
1974     55       SunPrairiecity     2        0.01           0.48   
1975     55         Superiorcity     8        0.03           0.34   
1976     55        TwoRiverscity     6        0.00           0.37   
1977     55        Watertowncity     2        0.00           0.52   
1978     55         Waukeshacity     6        0.08           0.45   
1979     55           Wausaucity     7        0.04           0.37   
1980     55        Wauwatosacity     9        0.06           0.35   
1981     55        WestAlliscity     6        0.09           0.29   
1982     55         WestBendcity     3        0.02           0.49   
1983     55  WhitefishBayvillage     9        0.01           0.42   
1984     55       Whitewatercity     6        0.00           0.65   
1985     55  WisconsinRapidscity     8        0.01           0.32   
1986     56         Cheyennecity     4        0.06           0.34   
1987     56         Evanstoncity     6        0.00           0.65   
1988     56         Gillettecity     9        0.01           0.53   
1989

## removing unpredictive columns from the dataset

In [13]:
del df2['state']
del df2['communityname']
del df2['fold']
del df2['ViolentCrimesPerPop']
del df2['highcrime']
df2.head()

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.01           0.61          0.21          0.83          0.02   
1        0.01           0.41          0.55          0.57          0.01   
2        0.03           0.34          0.86          0.30          0.04   
3        0.01           0.38          0.35          0.71          0.04   
4        0.04           0.37          0.32          0.70          0.21   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.01         0.41         0.49         0.26        0.21   
1         0.00         0.47         0.45         0.31        0.57   
2         0.01         0.41         0.42         0.27        0.59   
3         0.01         0.39         0.46         0.31        0.49   
4         0.02         1.00         1.00         1.00        0.14   

          ...           NumStreet  PctForeignBorn  PctBornSameState  \
0         ...                0.00            0.03              0.70   
1         ...                0.00            0.00              0.93   
2         ...                0.02            0.04              0.77   
3         ...                0.00            0.03              0.78   
4         ...                0.00            0.12              0.49   

   PctSameHouse85  PctSameCity85  PctSameState85  LandArea  PopDens  \
0            0.40           0.34            0.57      0.05     0.06   
1            0.66           0.82            0.84      0.11     0.03   
2            0.59           0.70            0.64      0.06     0.11   
3            0.56           0.67            0.71      0.09     0.05   
4            0.12           0.00            0.15      0.09     0.09   

   PctUsePubTrans  LemasPctOfficDrugUn  
0            0.01                  0.0  
1            0.01                  0.0  
2            0.04                  0.0  
3            0.00                  0.0  
4            0.01                  0.0  

[5 rows x 100 columns]

## Decision tree classifier

Use DecisionTreeClassifier to learn a decision tree to predict highCrime on the entire dataset (remember to exclude the crime rate feature from the input feature set so you are not cheating. 


In [14]:
from sklearn import tree
dtc = tree.DecisionTreeClassifier()
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import precision_recall_fscore_support





What are the training accuracy, precision, and recall for this tree? 

In [15]:
model1 = DecisionTreeClassifier()
model1.fit(df2,df1['highcrime'])
predict = model1.predict(df2)
print ("accuracy %2f" % model1.score(df2,df1['highcrime']))
print (precision_recall_fscore_support(df1['highcrime'],predict))




accuracy 1.000000
(array([ 1.,  1.]), array([ 1.,  1.]), array([ 1.,  1.]), array([ 743, 1250], dtype=int64))


ii.	What are the main features used for classification? Can you explain why they make sense (or not)?

In [231]:
features_columns = df2.columns
feature_importance = model1.feature_importances_
index = np.argsort(feature_importance)[::-1]
##feature_names = np.array(features_col_name)

print("Ranking of Features")
#print(top_features_index)
for i in range(10):
    print("{} feature no.{} ({}) ".format(i+1,features_columns[index[i]],feature_importance[index[i]]))


Ranking of Features
1 feature no.PctKids2Par (0.3619510290336175) 
2 feature no.racePctWhite (0.08646057909879695) 
3 feature no.racePctHisp (0.046550691242469855) 
4 feature no.PctEmplManu (0.026238998679621355) 
5 feature no.PctLess9thGrade (0.025989160493422314) 
6 feature no.MedRent (0.017598606496062884) 
7 feature no.PctWOFullPlumb (0.01500695232802416) 
8 feature no.PctIlleg (0.013286915203911736) 
9 feature no.PctPersOwnOccup (0.012873985895992554) 
10 feature no.PctEmploy (0.012015122428189327) 


What are the main features used for classification? Can you explain why they make sense (or not)?

The decision tree uses Gini index to determine the best attribute for splitting the dataset based on which the main feature for classification is determined.

The most important feature is PctKids2Par: This feature makes sense because kids with two parents are more vulnerable to criminal activities as both set of parents are not able to give complete attention due to reasons such as family issues and conflicting advices from both set of parents.
The other important features are racePctWhite, racePctHisp, PctEmplManu, PctLess9thGrade, MedRent


What are the 10-fold cross-validation accuracy, precision, and recall?

In [92]:
model1 = DecisionTreeClassifier()
print("Prec %f" % cross_val_score(model1, df2 , df1['highcrime'], cv=10, scoring = 'precision').mean() )
print("Recall %f" % cross_val_score(model1, df2 , df1['highcrime'], cv=10, scoring = 'recall').mean() )
print("Accuracy %f" % cross_val_score(model1, df2 , df1['highcrime'], cv=10, scoring = 'accuracy').mean() )


Prec 0.789863
Recall 0.786400
Accuracy 0.726997


Why are they different from the results in the previous test?

In previous test we used the same dataset for both training and testing which resulted into over-fitting of the model

However in this test we are using 10 fold cross-validation in calculating precession, recall and accuracy. 10 fold cross-validation consists of 10 iteration. Wherein each iteration we randomly select one fold for testing and the remaining nine folds for training. 

# Naive Bayes Classifier

What is the 10-fold cross-validation accuracy, precision, and recall for this method?

In [16]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import average_precision_score
model2 = GaussianNB()
print("Prec %f" % cross_val_score(model2 , df2, df1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(model2 , df2, df1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(model2 , df2, df1['highcrime'], cv=10, scoring='accuracy' ).mean())

Prec 0.911800
Recall 0.692000
Accuracy 0.761608


Prediction using Naive bayes

In [17]:
df3 = df2.copy()
model2.fit(df2,df1['highcrime'])
df3['predict'] = model2.predict(df2)
df3

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.01           0.61          0.21          0.83          0.02   
1           0.01           0.41          0.55          0.57          0.01   
2           0.03           0.34          0.86          0.30          0.04   
3           0.01           0.38          0.35          0.71          0.04   
4           0.04           0.37          0.32          0.70          0.21   
5           0.04           0.44          1.00          0.10          0.00   
6           0.41           0.37          1.00          0.02          0.03   
7           0.01           0.30          0.00          0.99          0.02   
8           0.00           0.39          0.31          0.75          0.02   
9           0.06           0.39          0.32          0.73          0.04   
10          0.07           0.41          0.53          0.57          0.05   
11          0.02           0.43          0.41          0.64          0.08   
12          0.01           0.46          0.67          0.47          0.02   
13          0.00           0.45          1.00          0.00          0.01   
14          0.04           0.33          0.33          0.73          0.02   
15          0.00           0.35          0.09          0.91          0.02   
16          0.05           0.32          0.55          0.56          0.04   
17          0.00           0.43          0.12          0.90          0.01   
18          0.02           0.23          0.16          0.85          0.08   
19          0.05           0.35          0.06          0.93          0.07   
20          0.01           0.44          0.26          0.80          0.01   
21          0.24           0.37          0.48          0.58          0.13   
22          0.00           0.54          0.34          0.71          0.07   
23          0.01           0.37          0.28          0.77          0.02   
24          0.01           0.36          0.15          0.83          0.16   
25          0.30           0.41          0.76          0.38          0.06   
26          0.28           0.45          0.83          0.34          0.04   
27          0.02           0.37          0.00          0.99          0.04   
28          0.01           0.44          0.44          0.64          0.05   
29          0.02           0.45          0.76          0.38          0.08   
...          ...            ...           ...           ...           ...   
1963        0.04           0.55          0.00          0.98          0.06   
1964        0.02           0.49          0.01          0.96          0.05   
1965        0.00           0.42          0.00          1.00          0.01   
1966        0.00           0.44          0.00          0.99          0.03   
1967        0.07           0.42          0.02          0.95          0.13   
1968        0.12           0.44          0.36          0.64          0.03   
1969        0.00           0.67          0.01          0.97          0.07   
1970        0.06           0.37          0.00          0.92          0.24   
1971        0.01           0.18          0.03          0.95          0.10   
1972        0.02           0.39          0.01          0.98          0.03   
1973        0.02           0.51          0.01          0.94          0.17   
1974        0.01           0.48          0.01          0.98          0.03   
1975        0.03           0.34          0.01          0.95          0.05   
1976        0.00           0.37          0.00          0.97          0.08   
1977        0.00           0.52          0.00          0.99          0.03   
1978        0.08           0.45          0.01          0.93          0.08   
1979        0.04           0.37          0.00          0.90          0.37   
1980        0.06           0.35          0.02          0.96          0.06   
1981        0.09           0.29          0.01          0.98          0.04   
1982        0.02           0.49          0.00          0.99          0.03   
19

 	What are the 10 most predictive features? This can be measured by the normalized absolute difference of means for the feature between the two classes:
|μ_T-μ_F |/(σ_T+σ_F )
The larger this different, the more predictive the feature. Why do these make sense (or not)?


In [18]:
from operator import itemgetter ,attrgetter,methodcaller

In [19]:
 df4 = df3[df3['predict'] == True]
 df5 = df3[df3['predict'] == False]  

 

In [20]:
resNb = []
for i in range(len(df2.columns)):
    resNb.append((abs(df4[[i]].mean() - df5[[i]].mean())  / df4[[i]].std() + df5[[i]].std()))

sorted(resNb,key = itemgetter(0),reverse = True)[:10]



[PctKids2Par    1.881312
 dtype: float64, pctWInvInc    1.841672
 dtype: float64, PctPersOwnOccup    1.769371
 dtype: float64, medFamInc    1.763879
 dtype: float64, PctFam2Par    1.757952
 dtype: float64, medIncome    1.733898
 dtype: float64, FemalePctDiv    1.679538
 dtype: float64, PctYoungKids2Par    1.59169
 dtype: float64, PctHousLess3BR    1.579829
 dtype: float64, TotalPctDiv    1.575829
 dtype: float64]

Why do these make sense (or not)?

We use Naïve Bayes Gaussian distribution for prediction. The important features of this distribution considered are Mean and Standard Deviation because in this distribution the data is normally distributed.

In this distribution the most predictive feature is determined by difference between the mean of true and false classes. We can observe that most predictive feature has the highest mean difference. Thus we can conclude that higher the difference between the mean of classification attributes the data is classified at a higher degree as the data is divided into purer sets resulting in improved classification   


How do these results compare with your results from decision trees, above?

Observation states that both decision tree classifier and NbGaussian classifier results in “PctKids2Par ”: population of children with two parents as the most important predictive feature with importance of  0.36 and 1.88 respectively.

We see that accuracy and precision for NB Gaussian classifier is finer than decision Tree classifier. This shows that for NB Gaussian classifier predictions provides true classification in most of the cases more than that of Decision Tree.

We also observe that; better the accuracy and  recall better is the classifier, as we need to know the true classes i.e. places with highest crime-rate. Using NB Gaussian classifier for the specified data-set, it can predict a low-crime area as a high-crime however it will never classify a high-crime area as a low-crime area.


# LinearSVC

What is the 10-fold cross-validation accuracy, precision, and recall for this method?

In [21]:
from sklearn.svm import LinearSVC
from sklearn.metrics import average_precision_score
model3 = LinearSVC()
print("Prec %f" % cross_val_score(model3 , df2, df1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(model3 , df2, df1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(model3 , df2, df1['highcrime'], cv=10, scoring='accuracy' ).mean())

Prec 0.845405
Recall 0.834400
Accuracy 0.796234


What are the 10 most predictive features? This can be measured by the absolute feature weights in the model. Why do these make sense (or not)?

In [22]:
import numpy as np
features_col_name = df2.columns
model3.fit(df2,df1['highcrime'])
feature_importance = model3.coef_[0]

top_features_index = np.argsort(feature_importance)[::-1]
feature_names = np.array(features_col_name)

print("Ranking of Features")
#print(top_features_index)
for i in range(10):
    print("{} feature no.{} ({}) ".format(i+1,features_col_name[top_features_index[i]],feature_importance[top_features_index[i]]))
#for ind in top_features_index:
#    print(feature_names[ind])

Ranking of Features
1 feature no.PersPerOccupHous (1.7551280762847878) 
2 feature no.RentHighQ (1.0668840915976965) 
3 feature no.MalePctDivorce (1.0656959142159945) 
4 feature no.NumUnderPov (1.0515453775737693) 
5 feature no.NumStreet (1.0191547807291406) 
6 feature no.PctOccupMgmtProf (1.0146759004251016) 
7 feature no.population (1.0023015096618844) 
8 feature no.agePct12t21 (0.9882726777322581) 
9 feature no.HousVacant (0.9365121945372911) 
10 feature no.agePct65up (0.8905731120431745) 


Based on the SVC the most important feature is 
PersPerOccupHous: mean persons per household (numeric - decimal)

We can see that when the number of people in a house is large there would be various issues such as money crisis resulting into feuds among family member. There would financial burden on the house head. It is very unlikely that a high percentage of large families are financially well-off  Also there would be lack of complete attention on kids in the household.  


How do these results compare with your results from decision trees, above?

Accuracy, precession and recall are higher for SVC than Decision Tree. SVC is more flexible towards penalties namely constraints or thresholds and loss functions which are used to scale large samples. Also SVC does not rely on entire training set data as compared to that of decision tree. Also decision tree’s predictions are based on over-fitted model. Thus it is proved based on the values. 

# Linear Model

Using 10-fold cross-validation, what is the estimated mean-squared-error (MSE) of the model?

In [24]:
from sklearn import linear_model
from sklearn import model_selection
import numpy as np
model4 = linear_model.LinearRegression()
model4.fit(df2, df1['ViolentCrimesPerPop'])
#print('mean squared error for regression: %f' %np.mean(model_selection.cross_val_score(regr, fil1, fil['highcrime'], 'mean_squared_error', cv=10)))
scores_mse = cross_val_score(model4, df2, df1['ViolentCrimesPerPop'], scoring = 'neg_mean_squared_error' ,cv=10)
scores_mse.mean()

-0.020093969304445319

MSE i.e. the mean squared error tells you how close a regression line is to a set of points. It does this by taking the distances from the points to the regression line (these distances are the “errors”) and squaring them. The squaring is necessary to remove any negative signs.

What is the MSE on the training set (train on all the data then test on it all)?

In [25]:
model4.fit(df2,df1['highcrime'])
predict = model4.predict(df2)

from sklearn.metrics import mean_squared_error
mean_squared_error(df1['ViolentCrimesPerPop'],predict)



0.21080313028166484

The MSE value is very low as we are testing and training on the same data-set which proves that Smaller the MSE, the closer you are to the regression line. Which justifies the results

What features are most predictive of a high crime rate?

In [26]:
features_col_name = df2.columns
feature_importance = model4.coef_

top_features_index = np.argsort(feature_importance)[::-1]
feature_names = np.array(features_col_name)

print("Ranking of Features")
#print(top_features_index)
for i in range(10):
    print("{}Top feature no.{} ({}) ".format(i+1,features_col_name[top_features_index[i]],feature_importance[top_features_index[i]]))

Ranking of Features
1Top feature no.population (3.9289306855757955) 
2Top feature no.PersPerOccupHous (1.205331889987419) 
3Top feature no.medFamInc (0.9268510944781458) 
4Top feature no.PctOccupMgmtProf (0.7439652169162789) 
5Top feature no.PctRecImmig8 (0.7123386069024439) 
6Top feature no.RentHighQ (0.6700684180174398) 
7Top feature no.MalePctDivorce (0.5991125827527165) 
8Top feature no.PctRecImmig5 (0.5856868766535908) 
9Top feature no.PctHousOwnOcc (0.5359550373145981) 
10Top feature no.PctForeignBorn (0.4463070159149707) 


What features are most predictive of a low crime rate?

In [27]:
top_features_index = np.argsort(feature_importance)
feature_names = np.array(features_col_name)

print("Ranking of Features")
#print(top_features_index)
for i in range(10):
    print("{}Top 10 low crime rate feature no.{} ({}) ".format(i+1,features_col_name[top_features_index[i]],feature_importance[top_features_index[i]]))

Ranking of Features
1Top 10 low crime rate feature no.numbUrban (-2.8305787286209596) 
2Top 10 low crime rate feature no.PctKids2Par (-1.2364633374179177) 
3Top 10 low crime rate feature no.PctRecImmig10 (-1.2130652747018975) 
4Top 10 low crime rate feature no.medIncome (-1.0115001213845476) 
5Top 10 low crime rate feature no.pctWInvInc (-0.9462543736498107) 
6Top 10 low crime rate feature no.NumIlleg (-0.847099741611971) 
7Top 10 low crime rate feature no.TotalPctDiv (-0.6513213678265983) 
8Top 10 low crime rate feature no.MedRent (-0.5834150739331968) 
9Top 10 low crime rate feature no.PctRecentImmig (-0.569406383359573) 
10Top 10 low crime rate feature no.PctBSorMore (-0.49266976635548965) 


We use coefficients of regression models for selecting and interpreting predictive features. This is based on the information that when all features are on the same scale, the most important feature should have the highest coefficient in the regression model, while features uncorrelated with the output variables should produce coefficient values close to zero. With simple regression models when the data is not very noisy this approach usually works

# RidgeCV

What is the estimated MSE of the model under 10-fold CV?

In [28]:
from sklearn.linear_model import RidgeCV
model5 = linear_model.RidgeCV(alphas = (10, 1, 0.1, 0.01, 0.001))
model5.fit(df2, df1['ViolentCrimesPerPop'])
mse = cross_val_score(model5,df2,df1['ViolentCrimesPerPop'], scoring = 'neg_mean_squared_error' ,cv=10)
mse.mean()

-0.019795021348168747

What is the MSE on the training set (train on all the data then test on it all)?

In [29]:
predict = model5.predict(df2)
mean_squared_error(df1['ViolentCrimesPerPop'],predict)


0.016763529155169436

What is the best alpha?

In [30]:
model5.alpha_

1.0

Alpha, generally known regularization; is used to set certain penalty so that the noise in the data-set can be minimized. Large values for alpha can result in regularization effect which dominates the squared loss function and the coefficients curves to zero. If the alpha value curves to zero and the solution inclines towards the ordinary least squares then large oscillations can be observed in coefficients. Hence, alpha should be selected such that balance is maintained between coefficients and penalty.

What does this say about the amount of overfitting in linear regression for this problem?


MSE is the major factor used in regression model determining whether the data is under-fitting or over-fitting the 
model. In linear regression as the we move towards next data points due to lack of biasness the data over-fits the 
model. However in Ridge regression, which is a regularized form of linear regression model which sets alpha as a 
penalty value which creates biasness preventing the model from over-fitting. 


# Polynomial Features

In [31]:
from sklearn.preprocessing import PolynomialFeatures

featureval = df2[list(df2)].values
labelval = df1['ViolentCrimesPerPop'].values
featureReg_x = np.array(featureval)
labelReg_y = np.array(labelval)

poly = PolynomialFeatures(degree=2)
X_ = poly.fit_transform(featureReg_x)



clf_poly = linear_model.LinearRegression()
clf_poly.fit(X_, labelReg_y)
predict_ = poly.fit_transform(featureReg_x)
y_pred = clf_poly.predict(predict_)
print(y_pred)


 

[ 0.06  0.14  1.   ...,  0.12  0.27  0.06]


What is the estimated MSE of the model under 10-fold CV?

In [32]:
mse = cross_val_score(clf_poly,df2,df1['ViolentCrimesPerPop'], scoring = 'neg_mean_squared_error' ,cv=10)
mse.mean()

-0.020093969304445319

What is the MSE on the training set (train on all the data then test on it all)?

In [33]:
mean_square_error = mean_squared_error(labelReg_y,y_pred)
mean_square_error

1.6958021953832663e-28

Does this mean the quadratic model is better than the linear model for this problem?

Calculating the difference between the MSE values for Quadratic model and Linear model we get minute result. Which states that the quadratic model and linear model for this dataset works in a similar fashion. This shows that the data points fall on a straight line rather than forming a curve. In case of a curved line quadratic model would have given better results.  


# Dirty data

Reading dirty data and based on the field ViolentCrimesPerPop making a traget class with True and False value

In [34]:
dirtydata = pd.read_csv("E:\Machine Learning\Crime Prediction Data2\Crime Prediction Data\communities-crime-full.csv")
feature_dirtydata = dirtydata.copy()
dirtydata['highcrime'] = dirtydata['ViolentCrimesPerPop'].map(lambda x : True if x >0.1 else False)



Removing unpredictive features from the dataset

In [35]:
del feature_dirtydata['state']
del feature_dirtydata['county']
del feature_dirtydata['community']
del feature_dirtydata['communityname']
del feature_dirtydata['fold']
del feature_dirtydata['ViolentCrimesPerPop']
feature_dirtydata

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
5           0.02           0.28          0.06          0.54          1.00   
6           0.01           0.39          0.00          0.98          0.06   
7           0.01           0.74          0.03          0.46          0.20   
8           0.03           0.34          0.20          0.84          0.02   
9           0.01           0.40          0.06          0.87          0.30   
10          0.13           0.71          0.15          0.07          1.00   
11          0.02           0.46          0.08          0.91          0.07   
12          0.03           0.47          0.01          0.96          0.13   
13          0.01           0.44          0.00          0.98          0.04   
14          0.04           0.36          0.01          0.85          0.14   
15          0.03           0.34          0.06          0.93          0.03   
16          0.15           0.31          0.40          0.63          0.14   
17          0.01           0.53          0.01          0.94          0.20   
18          0.02           0.47          0.01          0.97          0.07   
19          0.00           0.41          0.05          0.96          0.01   
20          0.25           0.54          0.05          0.71          0.48   
21          1.00           0.42          0.47          0.59          0.12   
22          0.01           0.34          0.02          0.87          0.07   
23          0.11           0.43          0.04          0.89          0.09   
24          0.02           0.96          0.05          0.00          1.00   
25          0.00           0.33          0.02          0.91          0.16   
26          0.06           0.49          0.46          0.00          1.00   
27          0.01           0.37          0.01          0.99          0.02   
28          0.01           0.27          0.43          0.64          0.08   
29          0.00           0.50          0.01          0.98          0.02   
...          ...            ...           ...           ...           ...   
1964        0.09           0.45          1.00          0.08          0.11   
1965        1.00           0.36          0.44          0.64          0.06   
1966        0.09           0.41          0.07          0.74          0.39   
1967        0.01           0.66          0.00          0.97          0.10   
1968        0.00           0.24          0.51          0.53          0.11   
1969        1.00           0.29          0.21          0.29          1.00   
1970        0.00           0.37          0.38          0.69          0.04   
1971        0.06           0.39          0.32          0.73          0.04   
1972        0.08           0.49          0.07          0.69          0.27   
1973        0.00           0.38          0.03          0.97          0.04   
1974        0.00           0.41          0.62          0.51          0.02   
1975        0.05           0.27          0.03          0.89          0.27   
1976        0.06           0.29          0.03          0.93          0.18   
1977        0.01           0.35          0.06          0.94          0.06   
1978        0.01           0.54          0.02          0.94          0.13   
1979        0.02           0.43          0.39          0.63          0.15   
1980        0.06           0.43          0.04          0.93          0.03   
1981        0.07           0.38          0.17          0.84          0.11   
1982        0.00           0.47          0.02          0.98          0.03   
1983        0.01           0.45          0.37          0.65          0.15   
19

Replacing ? with the mean of the repsective rows

In [36]:
for i in feature_dirtydata.columns:
    feature_dirtydata[[i]] = feature_dirtydata[[i]].convert_objects(convert_numeric=True)
    feature_dirtydata[[i]]= feature_dirtydata[[i]].fillna(feature_dirtydata[[i]].mean())
feature_dirtydata

C:\Users\sunvi\Anaconda3\lib\site-packages\ipykernel\__main__.py:2: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  from ipykernel import kernelapp as app


population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
5           0.02           0.28          0.06          0.54          1.00   
6           0.01           0.39          0.00          0.98          0.06   
7           0.01           0.74          0.03          0.46          0.20   
8           0.03           0.34          0.20          0.84          0.02   
9           0.01           0.40          0.06          0.87          0.30   
10          0.13           0.71          0.15          0.07          1.00   
11          0.02           0.46          0.08          0.91          0.07   
12          0.03           0.47          0.01          0.96          0.13   
13          0.01           0.44          0.00          0.98          0.04   
14          0.04           0.36          0.01          0.85          0.14   
15          0.03           0.34          0.06          0.93          0.03   
16          0.15           0.31          0.40          0.63          0.14   
17          0.01           0.53          0.01          0.94          0.20   
18          0.02           0.47          0.01          0.97          0.07   
19          0.00           0.41          0.05          0.96          0.01   
20          0.25           0.54          0.05          0.71          0.48   
21          1.00           0.42          0.47          0.59          0.12   
22          0.01           0.34          0.02          0.87          0.07   
23          0.11           0.43          0.04          0.89          0.09   
24          0.02           0.96          0.05          0.00          1.00   
25          0.00           0.33          0.02          0.91          0.16   
26          0.06           0.49          0.46          0.00          1.00   
27          0.01           0.37          0.01          0.99          0.02   
28          0.01           0.27          0.43          0.64          0.08   
29          0.00           0.50          0.01          0.98          0.02   
...          ...            ...           ...           ...           ...   
1964        0.09           0.45          1.00          0.08          0.11   
1965        1.00           0.36          0.44          0.64          0.06   
1966        0.09           0.41          0.07          0.74          0.39   
1967        0.01           0.66          0.00          0.97          0.10   
1968        0.00           0.24          0.51          0.53          0.11   
1969        1.00           0.29          0.21          0.29          1.00   
1970        0.00           0.37          0.38          0.69          0.04   
1971        0.06           0.39          0.32          0.73          0.04   
1972        0.08           0.49          0.07          0.69          0.27   
1973        0.00           0.38          0.03          0.97          0.04   
1974        0.00           0.41          0.62          0.51          0.02   
1975        0.05           0.27          0.03          0.89          0.27   
1976        0.06           0.29          0.03          0.93          0.18   
1977        0.01           0.35          0.06          0.94          0.06   
1978        0.01           0.54          0.02          0.94          0.13   
1979        0.02           0.43          0.39          0.63          0.15   
1980        0.06           0.43          0.04          0.93          0.03   
1981        0.07           0.38          0.17          0.84          0.11   
1982        0.00           0.47          0.02          0.98          0.03   
1983        0.01           0.45          0.37          0.65          0.15   
19

Precision Reacll and accuracy in dirty data using cross fold

In [37]:
model_dirty = DecisionTreeClassifier()
print("Prec %f" % cross_val_score(model_dirty, feature_dirtydata , dirtydata['highcrime'], cv=10, scoring = 'precision').mean() )
print("Recall %f" % cross_val_score(model_dirty, feature_dirtydata , dirtydata['highcrime'], cv=10, scoring = 'recall').mean() )
print("Accuracy %f" % cross_val_score(model_dirty, feature_dirtydata , dirtydata['highcrime'], cv=10, scoring = 'accuracy').mean() )

Prec 0.817103
Recall 0.795333
Accuracy 0.762750


Precision recall and accuracy values when training and testing is performed on same dataset

In [38]:
model_dirty.fit(feature_dirtydata,dirtydata['highcrime'])
predict = model_dirty.predict(feature_dirtydata)
print (precision_recall_fscore_support(dirtydata['highcrime'],predict))


(array([ 1.,  1.]), array([ 1.,  1.]), array([ 1.,  1.]), array([ 743, 1251], dtype=int64))


Important features of DecisionTree classifier on dirty data

In [39]:
features_columns = feature_dirtydata.columns
feature_importance = model_dirty.feature_importances_
index = np.argsort(feature_importance)[::-1]
##feature_names = np.array(features_col_name)

print("Ranking of Features")
#print(top_features_index)
for i in range(10):
    print("{} feature no.{} ({}) ".format(i+1,features_columns[index[i]],feature_importance[index[i]]))

Ranking of Features
1 feature no.PctKids2Par (0.36224029458015017) 
2 feature no.racePctWhite (0.08648036391601062) 
3 feature no.racePctHisp (0.05000104731332353) 
4 feature no.PctEmplManu (0.026367034338736432) 
5 feature no.PctLess9thGrade (0.02066687029035681) 
6 feature no.blackPerCap (0.018214142789072345) 
7 feature no.MedRent (0.015097721105857083) 
8 feature no.PctImmigRec10 (0.013829526881682312) 
9 feature no.PctEmploy (0.013360189756275469) 
10 feature no.PctSpeakEnglOnly (0.01293737156460395) 


What does this say about the effect of missing values?

The current results, after imputing the missing values with the mean, are better than the previous results. One possible reason being previously we deleted the entire column having missing values however later we are imputing the missing values with the mean of the corresponding column and the previously deleted column is taken into consideration for computing the prediction. Thus the newly added columns after imputing might have been playing important role in predictions. 

# 5a. perceptron

In [40]:
from sklearn.linear_model import Perceptron
model8 = Perceptron()
model8.fit(df2,df1['highcrime'])
print("Prec %f" % cross_val_score(model8 , df2, df1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(model8 , df2, df1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(model8 , df2, df1['highcrime'], cv=10, scoring='accuracy' ).mean())

Prec 0.879222
Recall 0.779200
Accuracy 0.787683


In [41]:
features_col_name = df2.columns
feature_importance = model8.coef_[0]

top_features_index = np.argsort(feature_importance)[::-1]
feature_names = np.array(features_col_name)

print("Ranking of Features")
#print(top_features_index)
for i in range(10):
    print("{}Top feature no.{} ({}) ".format(i+1,features_col_name[top_features_index[i]],feature_importance[top_features_index[i]]))

Ranking of Features
1Top feature no.racepctblack (16.71999999999997) 
2Top feature no.racePctHisp (12.210000000000003) 
3Top feature no.PctPersDenseHous (10.07000000000001) 
4Top feature no.MalePctDivorce (9.340000000000018) 
5Top feature no.PctIlleg (9.22) 
6Top feature no.OtherPerCap (8.219999999999992) 
7Top feature no.TotalPctDiv (7.679999999999988) 
8Top feature no.FemalePctDiv (6.970000000000013) 
9Top feature no.RentHighQ (6.960000000000011) 
10Top feature no.HousVacant (6.779999999999991) 


In [42]:
model8.fit(feature_dirtydata,dirtydata['highcrime'])
print("Prec %f" % cross_val_score(model8, feature_dirtydata , dirtydata['highcrime'], cv=10, scoring = 'precision').mean() )
print("Recall %f" % cross_val_score(model8, feature_dirtydata , dirtydata['highcrime'], cv=10, scoring = 'recall').mean() )
print("Accuracy %f" % cross_val_score(model8, feature_dirtydata , dirtydata['highcrime'], cv=10, scoring = 'accuracy').mean() )

Prec 0.757628
Recall 0.931200
Accuracy 0.750300


In [43]:
features_col_name = feature_dirtydata.columns
feature_importance = model8.coef_[0]

top_features_index = np.argsort(feature_importance)[::-1]
feature_names = np.array(features_col_name)

print("Ranking of Features")
#print(top_features_index)
for i in range(10):
    print("{}Top feature no.{} ({}) ".format(i+1,features_col_name[top_features_index[i]],feature_importance[top_features_index[i]]))

Ranking of Features
1Top feature no.racepctblack (23.949999999999953) 
2Top feature no.racePctHisp (10.400000000000006) 
3Top feature no.OtherPerCap (9.559999999999999) 
4Top feature no.PctPersDenseHous (9.210000000000022) 
5Top feature no.HousVacant (8.880000000000019) 
6Top feature no.PctIlleg (8.709999999999985) 
7Top feature no.PctSameCity85 (8.330000000000009) 
8Top feature no.RentHighQ (8.12000000000002) 
9Top feature no.MalePctDivorce (7.869999999999992) 
10Top feature no.PopDens (7.279999999999992) 


In [44]:
from sklearn.ensemble import RandomForestClassifier

In [45]:
model9 = RandomForestClassifier(n_jobs=2)
model9.fit(df2, df1['highcrime'])


print ("Recall %f" %cross_val_score(model9 , df2, df1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Precision %f" %cross_val_score(model9 , df2, df1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Accuracy %f" %cross_val_score(model9 ,df2, df1['highcrime'], cv=10, scoring='accuracy' ).mean())

from sklearn.datasets import make_classification
from sklearn.ensemble import ExtraTreesClassifier

# Build a classification task using 3 informative features

total_feature = df2.columns
important_feature = model9.feature_importances_
import numpy as np
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
indices
print ('Feature Ranking') 
for i in range(10):
    print ("{} Top 10 high feature are{} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

Recall 0.821600
Precision 0.844691
Accuracy 0.781168
Feature Ranking
1 Top 10 high feature arePctIlleg (0.06979629798028167)
2 Top 10 high feature areTotalPctDiv (0.04433942734902356)
3 Top 10 high feature arePctFam2Par (0.043480668954438764)
4 Top 10 high feature areracePctWhite (0.04107797758448359)
5 Top 10 high feature arePctKids2Par (0.03940080965875138)
6 Top 10 high feature arePctPopUnderPov (0.037048719574586116)
7 Top 10 high feature arepctWPubAsst (0.03651323984004249)
8 Top 10 high feature arepctWInvInc (0.03529391630813689)
9 Top 10 high feature areMalePctDivorce (0.03334648584903445)
10 Top 10 high feature arePctPersOwnOccup (0.029057967196945762)


In [224]:
model9 = RandomForestClassifier(n_jobs=2)
model9.fit(feature_dirtydata, dirtydata['highcrime'])


print ("Recall %f" %cross_val_score(model9 , feature_dirtydata, dirtydata['highcrime'], cv=10, scoring='recall' ).mean())
print ("Precision %f" %cross_val_score(model9 ,feature_dirtydata, dirtydata['highcrime'], cv=10, scoring='precision' ).mean())
print ("Accuracy %f" %cross_val_score(model9 ,feature_dirtydata, dirtydata['highcrime'], cv=10, scoring='accuracy' ).mean())

from sklearn.datasets import make_classification
from sklearn.ensemble import ExtraTreesClassifier

# Build a classification task using 3 informative features

total_feature = feature_dirtydata.columns
important_feature = model9.feature_importances_
import numpy as np
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
indices
print ('Feature Ranking') 
for i in range(10):
    print ("{} Top 10 high feature are{} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

Recall 0.829733
Precision 0.875660
Accuracy 0.812409
Feature Ranking
1 Top 10 high feature areFemalePctDiv (0.07595384725791428)
2 Top 10 high feature arePctKids2Par (0.0606900460681283)
3 Top 10 high feature areracePctWhite (0.05165891292521445)
4 Top 10 high feature areNumIlleg (0.04406528615734501)
5 Top 10 high feature arepctWInvInc (0.04251125032649126)
6 Top 10 high feature arePctPersDenseHous (0.04099100377976288)
7 Top 10 high feature areracepctblack (0.0389967604639494)
8 Top 10 high feature arePctIlleg (0.03010664256058086)
9 Top 10 high feature areHousVacant (0.026736824628999168)
10 Top 10 high feature areTotalPctDiv (0.02462737633989869)


Accuracy 0.787683 for clean dataset for perceptron
Accuracy 0.781168 for clean dataset for randomforest

Accuracy 0.750300 for dirty dataset for perceptron
Accuracy 0.812409 for dirty dataset for randomforest